### ISSUE TITLE
SuperH: How to run processor emulation test framework?

### ISSUE THREAD

**Issue by VGKintsugi** [NationalSecurityAgency/ghidra#1077]
I would like to run the built in emulation test framework against the SuperH SH-1/SH-2/SH-2A processor module. emteere mentioned it here: https://github.com/NationalSecurityAgency/ghidra/pull/715#issuecomment-512566002

Are there any instructions on how to use the emulation test framework? I don't even know which binary to launch or if it's built into Ghidra itself. Thanks in advance.

**Comment by mumbel** [comment]
I actually tried doing this yesterday for superh 2, but failed to get a working gcc, so gave up in defeat.

You will need to get `sh-elf-gcc`, `sh-elf-nm`, `sh-elf-objdump`, and `sh-elf-readelf` and the lib needed to build an elf file (basically libgcc i think).  This is your `--toolchain-root` argument.   The script is semi-broken as `toolchain-root` and a few others values are not assigned from parseargs. (I made a few adjustments in a branch for my testing  https://github.com/mumbel/ghidra/tree/pcode/Ghidra/Extensions/SleighDevTools/pcodetest) 

Add your `SH2` section to `SleighDevTools\pcodetests\pcode_defs.py`.

Run the `SleighDevTools\pcodetests\build` script on the `--pcodetest=SH2` with whatever other args are needed, probably just toolchain-root and build-root.  It should successfully build a `O0` and a `O3` build-* directory with all the intermediate C files and output files.  You really just care about the *.out files.  (this is where I got hung up as when I built my cross-compiler I didn't end up with the additional binaries to build a linked elf, and several symbols were missing).

Once you have the *.out, move the *.out to `Ghidra\Processors\SuperH\data\pcodetests\`

In eclipse:

You'll need to then create the emulator tests.  I basically just copied tests from the other processors and renamed the files/classes/strings.  I would add a `SuperH/src/test.processors/java` directory and then add a java package `ghidra.test.processors` to that directory.  Then add java classes like:  `SH2_O0_EmulatorTest.java` and `SH2_O3_EmulatorTest.java` to that package that extend `ProcessorEmulatorTestAdapter`.  These are pretty minimal files once you look at the contents of the other tests.

You can then right click the .java and run as a junit test

This will run through each .TEST from https://github.com/NationalSecurityAgency/ghidra/tree/master/Ghidra/Extensions/SleighDevTools/pcodetest/c_src

You should see it load the .out and analyze it headlessly, then it will start processing, and you should see a number of passes and errors of a total number asserts.  From the Junit view you can see which tests passed or failed as a whole.  you can then individually run by right clicking and running from that view.

**Comment by VGKintsugi** [comment]
I was able to get SH2 gcc compiled by following these instructions: https://github.com/NationalSecurityAgency/ghidra/pull/715#issuecomment-512900540. 

I couldn't get the --toolchain-root so I ended up creating a /local/Toolchains/SuperH/ directory. I copied the bins and and cc1 to there as well. Eventually I got the compilation process to start, but it failed out with: 

/tmp/ccYHT4VW.o: In function `biopDividu1u1':
BIOPS2_BODY.c:(.text+0x110c): undefined reference to `__udivsi3_i4i'
/tmp/ccYHT4VW.o: In function `biopDividu2u2':
BIOPS2_BODY.c:(.text+0x116c): undefined reference to `__sdivsi3_i4i'
/tmp/ccYHT4VW.o: In function `biopDividu4u4':
BIOPS2_BODY.c:(.text+0x11c0): undefined reference to `__udivsi3_i4i'
/tmp/ccYHT4VW.o: In function `biopRemainderu1u1':
BIOPS2_BODY.c:(.text+0x1230): undefined reference to `__udivsi3_i4i'
/tmp/ccYHT4VW.o: In function `biopRemainderu2u2':
BIOPS2_BODY.c:(.text+0x12a0): undefined reference to `__udivsi3_i4i'
/tmp/ccYHT4VW.o: In function `biopRemainderu4u4':
BIOPS2_BODY.c:(.text+0x1304): undefined reference to `__udivsi3_i4i'
/tmp/ccUBAaCU.o: In function `biopDividi4i4':
BIOPS_BODY.c:(.text+0x117c): undefined reference to `__sdivsi3_i4i'
/tmp/ccUBAaCU.o: In function `biopRemainderi4i4':
BIOPS_BODY.c:(.text+0x11e0): undefined reference to `__sdivsi3_i4i'
/tmp/ccUBAaCU.o: In function `biopRemainderi2i2':
BIOPS_BODY.c:(.text+0x1430): undefined reference to `__sdivsi3_i4i'
/tmp/ccUBAaCU.o: In function `biopDividi2i2':
BIOPS_BODY.c:(.text+0x152c): undefined reference to `__sdivsi3_i4i'
/tmp/ccUBAaCU.o: In function `biopRemainderi1i1':
BIOPS_BODY.c:(.text+0x16ac): undefined reference to `__sdivsi3_i4i'
/tmp/ccUBAaCU.o:BIOPS_BODY.c:(.text+0x1758): more undefined references to `__sdivsi3_i4i' follow
collect2: error: ld returned 1 exit status

# WARNING: SH2_GCC_O0_pcodetest.out does not exist
# ERROR: output not created SH2_GCC_O0_pcodetest.out

Will keep beating at this.

**Comment by mumbel** [comment]
Sweet, that looks better than I had, I also had shifting symbols it couldn't find.  Divide seemed like it was gonna be a trouble point just since that has its own gcc flag.  Did any of the lib/ files built in cross-compiler end up with that symbol.  Did you end up with a similar pcode_defs to SuperH4 (using the `-L`)?

I also had the thought about `#ifdef` for multiply and divide while I was trying this out and have the same functionality as HAS_FLOAT, HAS_DOUBLE, and HAS_LONGLONG.

**Comment by GhidorahRex** [comment]
What gcc flags do you have for the pcodetest for SH2?

**Comment by VGKintsugi** [comment]
I tried with the SH4 command line: 'ccflags': '-mb -mrenesas -m4 -L %(toolchain_dir)s/lib/gcc/sh4-elf/%(gcc_version)s -lgcc',

I switched the "-m4" to "-m2"

**Comment by VGKintsugi** [comment]
It was a pain but I got SH2_GCC_O0_pcodetest.out and SH2_GCC_O3_pcodetest.out to build. After building the Saturn-SDK-GCC-SH2 tool chain, I made a SuperH/saturn-sh2-elf directory in the "toolchain" directory. I copied the contents of Saturn-SDK-GCC-SH2/toolchain into the Saturn-SDK-GCC-SH2/toolchain/SuperH/saturn-sh2-elf directory. 

In SuperH/saturn-sh2-elf/bin I made symlinks for gcc, nm, objdump, and readelf and pointed them to their saturn-sh2-elf-* equivalents. I then edited PCodeTest.defaults.toolchain_root to point to my /Saturn-SDK-GCC-SH2/toolchain/ directory. 

I'm now trying to figure out how to run the tests. It's unclear what project I need to open in Eclipse. Do I need to build Ghidra itself from source?
